src/librbd/mirror/snapshot/GroupImageCreatePrimaryRequest.cc

```diff
@@ -0,0 +1,395 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/mirror/snapshot/GroupImageCreatePrimaryRequest.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "librbd/ExclusiveLock.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/ImageState.h"
+#include "librbd/ImageWatcher.h"
+#include "librbd/Operations.h"
+#include "librbd/Utils.h"
+#include "librbd/mirror/snapshot/Utils.h"
+
+#include <shared_mutex> // for std::shared_lock
+
+#define dout_subsys ceph_subsys_rbd
+
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::mirror::snapshot::GroupImageCreatePrimaryRequest: " \
+                           << this << " " << __func__ << ": "
+
+namespace librbd {
+namespace mirror {
+namespace snapshot {
+
+using librbd::util::create_context_callback;
+using librbd::util::create_rados_callback;
+using librbd::util::snap_create_flags_api_to_internal;
+using librbd::util::get_default_snap_create_flags;
+template <typename I>
+GroupImageCreatePrimaryRequest<I>::GroupImageCreatePrimaryRequest(
+    std::vector<I *>&image_ctxs, std::vector<std::string> &global_image_ids,
+    std::vector<uint64_t> &clean_since_snap_ids,
+    uint64_t group_snap_create_flags, uint32_t flags,
+    const std::string &group_snap_id, std::vector<uint64_t>&snap_ids,
+    Context *on_finish)
+  : m_image_ctxs(image_ctxs), m_global_image_ids(global_image_ids),
+    m_clean_since_snap_ids(clean_since_snap_ids),
+    m_group_snap_create_flags(group_snap_create_flags), m_flags(flags),
+    m_group_snap_id(group_snap_id), m_snap_ids(snap_ids),
+    m_on_finish(on_finish) {
+
+  if (m_group_snap_id.empty() && m_image_ctxs.size() > 1) {
+    finish(-EINVAL);
+    return;
+  }
+
+  m_cct = m_image_ctxs[0]->cct;
+
+  //FIXME: using the first image_ctx pool for now
```

Comment:
We are already rewriting CreatePrimaryRequest.cc to fit to group mirroring requirements. Many routines needs a rewrite including the way we get mirror peers for different images when we want cross pool images in the group (its again a rewrite). Rather I would suggest keep it generic enough, by using different image pool and their respective peers.

Reply:
Fixed